controller/src/main/java/io/pravega/controller/store/stream/records/EpochRecord.java

```diff
@@ -50,14 +50,19 @@
     private final Map<Long, StreamSegmentRecord> segmentMap = new HashMap<>();
     
     @Builder
-    EpochRecord(int epoch, int referenceEpoch, List<StreamSegmentRecord> segments, long creationTime) {
+    private EpochRecord(int epoch, int referenceEpoch, List<StreamSegmentRecord> segments, long creationTime, boolean copyCollections) {
         this.epoch = epoch;
         this.referenceEpoch = referenceEpoch;
-        this.segments = ImmutableList.copyOf(segments);
+        this.segments = copyCollections ? ImmutableList.copyOf(segments) : segments;
```

Comment:
This looks like bug prone. `EpochRecord` is a data class, right? As such, don't we expect the state of its objects to be immutable?

Reply:
Yes, its a data class. But someone constructing it gives it a collection. So we used to make an immutable copy of the collection by default.
When deserializing, the serializer gives us a collection that we know for sure will not be mutated because the deserializer is owned within this data class. so we avoid a duplicate copy of collection here by using this copy collections flag.

Reply:
This becomes more relevant for streams with very large number of concurrent segments. 
For a stream with 10k segments (like we create in metadata scalability system tests), this avoids creation of 10k duplication collection node for each such object. 

A rough calculation where 32 bytes per object (16 for the object and 16 for the reference fields) will translate to ~320kb of raw collection node footprint. The java object with its paddings and alignments takes more. Maybe around 500kb or 0.5Mb. This is just arising from creating a copy of the collection which can be avoided altogether. 

For each such object that we create, we could have large number of such collection nodes that will have a memory footprint until garbage collected.

Reply:
employed the approach to use immutable collections in our serializers.